core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java

```diff
@@ -303,6 +304,9 @@ public void serverStopping(Server server) {
             public void serverStopped(Server server) throws Exception {
                 if (healthCheckerListener != null) {
                     healthCheckers.stream().map(ListenableHealthChecker.class::cast).forEach(c -> {
+                        if (c instanceof SafeCloseable) {
+                            ((SafeCloseable) c).close();
```

Comment:
What would happen if a user restarts a server? :thinking: It seems to me that it's much easier to deal with the life cycle issues like this by triggering one-time task when previously cached healthiness expires.

Reply:
You mean manually restart in the same process? Right I didn't aware starting the same server which won't re-create the instance.

I'm curious about how to(or who) trigger after expired? Especially if we are using the long polling, will next checking far longer than expected?

Reply:
> Especially if we are using the long polling, will next checking far longer than expected?

That's a good point I missed. How about letting a user specify a TTL for each check? e.g.

```java
HealthChecker.of(Supplier<? extends CompletionStage<HealthCheckStatus>> supplier) { ... }

public final class HealthCheckStatus {
    // The current healthiness
    private boolean isHealthy;
    // When we need to check again
    private long ttlMillis;
}
```

and then, we could call `supplier.get()` whenever we need to update the cached status, and stop calling them when the server is stopped. (and cancel the future returned by `supplier.get()`?)

I guess then the health checker implementation will need to be tightly coupled with `HealthCheckService` although it should be hidden from users.

What do you think?

Reply:
SGTM, for confirmation, do you mean put the EventExecutor at `HealthCheckService` and make it can call the supplier from `HealthChecker`? If I'm not wrong.

Reply:
Something like https://github.com/kojilin/armeria/compare/3392...kojilin:3392-2?expand=1 ?

Reply:
Something like that, but I guess we can reuse the current health check request's `ServiceRequestContext.eventLoop().withoutContext()`, rather than using `CommonPools`?